src/mimeparser.rs

```diff
@@ -376,6 +376,12 @@ impl MimeMessage {
         if !encrypted {
             signatures.clear();
         }
+        if let Some(peerstate) = decryption_info.peerstate.as_mut().filter(|peerstate| {
+            peerstate.prefer_encrypt != EncryptPreference::Mutual && !signatures.is_empty()
+        }) {
+            peerstate.prefer_encrypt = EncryptPreference::Mutual;
+            peerstate.save_to_db(&context.sql).await?;
+        }
```

Comment:
Nit: I think that usually it's better to avoid more rare things like `as_mut()` and `filter()` (as long as it makes sense), and use simpler concepts like a nested `if`, because that makes it easier for less experienced people to read the code. And, personally, I would also be a bit faster to read a nested `if` than the code as it is right now.

Bottom line: I would personally prefer the nested `if` version, but I'm fine with the code as-is, too.

```rust
        if !signatures.is_empty() {
            if let Some(peerstate) = &mut decryption_info.peerstate {
                if peerstate.prefer_encrypt != EncryptPreference::Mutual {
                    peerstate.prefer_encrypt = EncryptPreference::Mutual;
                    peerstate.save_to_db(&context.sql).await?;
                }
            }
        }
```

or 
```rust
        if let Some(peerstate) = &mut decryption_info.peerstate {
            if !signatures.is_empty() && peerstate.prefer_encrypt != EncryptPreference::Mutual {
                peerstate.prefer_encrypt = EncryptPreference::Mutual;
                peerstate.save_to_db(&context.sql).await?;
            }
        }
```